verification/spec/relayer/environment.tla

```diff
@@ -0,0 +1,275 @@
+                ]
+         ] IN
+     
+    \* update chain with "ClientUpdate" datagrams
+    LET updateClients == {dgr \in datagrams : dgr.type = "ClientUpdate"} IN
+    LET updateClientIDs == {clientID \in ClientIDs : 
+            /\ \E dgr \in updateClients : dgr.clientID = clientID
+            /\ ValidClientID(chainID, clientID)} IN
+    LET updateClientHeights(clientID) == 
+            {h \in Heights : \E dgr \in updateClients : 
+                dgr.clientID = clientID /\ dgr.height = h} IN  
+    
+    \* new chain record with clients updated
+    LET clientsUpdated == [
+            height |-> clientsCreated.height,
+            clientHeights |-> 
+                [cID \in ChainIDs |-> IF GetClientID(chainID, cID) \in updateClientIDs
+                                      THEN Max(updateClientHeights(GetClientID(chainID, cID)))
+                                      ELSE clientsCreated.clientHeights[cID] 
+                ]
+         ] IN
+    
+    clientsUpdated
+ 
+\* Update chainID with the received datagrams
+\* Currently, only supporting ICS2: Client updates
+\* TODO: Support the remaining datagrams
+UpdateChain(chainID, datagrams) == 
+    LET chain == chains[chainID] IN
+    LET lightClientsUpdated == LightClientUpdate(chainID, chain, datagrams) IN 
+    
+    lightClientsUpdated
+    
+(***************
+Chain actions
+***************)       
+
+\* Advance the height of the chain until MaxHeight is reached
+AdvanceChain ==
+    \E chainID \in ChainIDs :
+        /\ chains[chainID].height < MaxHeight
+        /\ chains' = [chains EXCEPT 
+                        ![chainID].height = chains[chainID].height + 1                                            
+                     ]
+        /\ UNCHANGED pendingDatagrams
+
+\* Update the height of the light client on chainID for chainID 
+UpdateOwnClient ==
+    \E chainID \in ChainIDs :
+        /\ chains[chainID].clientHeights[chainID] < chains[chainID].height
+        /\ chains' = [chains EXCEPT 
+                        ![chainID].clientHeights = 
+                            [chains[chainID].clientHeights EXCEPT
+                                ![chainID] = chains[chainID].height 
+                            ]
+                     ]
+        /\ UNCHANGED pendingDatagrams
+
+\* Receive the datagrams and update the chain state        
+ReceiveDatagrams ==
+    \E chainID \in ChainIDs :
+        /\ pendingDatagrams[chainID] /= {} 
+        /\ chains' = [chains EXCEPT 
+                        ![chainID] = UpdateChain(chainID, pendingDatagrams[chainID])
+                     ]
+                        
+        /\ pendingDatagrams' = [pendingDatagrams EXCEPT
+                                    ![chainID] = {}
+                               ]
+\* Initial value if the chain: 
+\*      - height is initialized to 1
+\*      - all the light clients are uninitialized
+InitChain ==
+    [height |-> 1,
+     clientHeights |-> [chID \in ChainIDs |-> nullHeight]]
```

Comment:
I am wondering if it would be simpler (easier to understand) if we just consider two chains (Alice and Bob) and then chain state is height and lightClient of other chain.